pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/RRDD.scala

```diff
@@ -27,21 +30,32 @@ private abstract class BaseRRDD[T: ClassTag, U: ClassTag](
 
   override def compute(split: Partition, context: TaskContext): Iterator[U] = {
 
-    val parentIterator = firstParent[T].iterator(split, context)
+    val serverSocket = new ServerSocket(0, 2)
+    val listenPort = serverSocket.getLocalPort()
 
-    val pb = rWorkerProcessBuilder()
+    val pb = rWorkerProcessBuilder(listenPort)
+    pb.redirectErrorStream()  // redirect stderr into stdout
     val proc = pb.start()
+    val errThread =  startStdoutThread(proc)
 
-    val errThread = startStderrThread(proc)
+    // We use two socket ot separate input and output, then it's easy to manage
+    // the lifecycle of them to avoid deadlock.
+    // TODO: optimize it to use one socket
+
+    // the socket used to send out the input of task
+    serverSocket.setSoTimeout(10000)
+    val inSocket = serverSocket.accept()
+    val parentIterator = firstParent[T].iterator(split, context)
```

Comment:
So one reason the `parentIterator` call was at the beginning is that if two RRDDs were chained then the child R process's iterator will launch R do some computation and only when its over will the parent R process start.
This was mainly to avoid having too many R processes open at the same time -- Can we push this line back to the top of this method ?